### PR TITLE
Add Redis Redisson support

### PR DESCRIPTION
**Pull request by dengliming**
## What does this PR do?
A PR for [https://github.com/elastic/apm-agent-java/issues/1054](https://github.com/elastic/apm-agent-java/issues/1054)

## Checklist

- [x] My code follows the [style guidelines of this project](CONTRIBUTING.md#java-language-formatting-guidelines)
- [x] I have rebased my changes on top of the latest master branch

- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#testing) pass locally with my changes

- [x] I have updated [CHANGELOG.asciidoc](CHANGELOG.asciidoc)
- [x] I have updated [supported-technologies.asciidoc](docs/supported-technologies.asciidoc)
- [x] Added an instrumentation plugin? How did you make sure that old, non-supported versions are not instrumented by accident?

## Related issues

 [https://github.com/elastic/apm-agent-java/issues/1054](https://github.com/elastic/apm-agent-java/issues/1054)
## Screenshots
<!-- _(Optional)_
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.

### CODE REVIEW

File: apm-agent-plugins/apm-redis-plugin/apm-redisson-plugin/src/test/java/co/elastic/apm/agent/redis/redisson/RedissonInstrumentationTest.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        redisson = Redisson.create(config);
+    }
+
+    @AfterEach
+    void shutdown() {
+        redisson.shutdown();
+    }
+
+    @Test
+    void testRedission() {
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-redis-plugin/apm-redisson-plugin/src/test/java/co/elastic/apm/agent/redis/redisson/RedissonInstrumentationTest.java]: Are there other ways of using the Redisson client? If so, please add a test for each API usage.

**Comment by dengliming** [comment]
> @dengliming Thanks for this awesome contribution! ❤️
> Please see a few comments inside.
> Besides, you must sign the [contributor agreement](https://www.elastic.co/contributor-agreement), otherwise we will not be able to merge this PR.

I've signed it step by step, but it still shows this

File: apm-agent-plugins/apm-redis-plugin/apm-redisson-plugin/pom.xml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>test-jar</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
```
[the remark this answers is not in the document]
Comment (dengliming) [reply apm-agent-plugins/apm-redis-plugin/apm-redisson-plugin/pom.xml]: It's useless. Let me remove it

**Comment by dengliming** [comment]
@eyalkoren Please review these changes. Thanks.

**Comment by eyalkoren** [comment]
@dengliming Thanks for the quick turnaround!
Looks great!
Please make sure you can build the agent using `./mvnw clean package` properly (with JDK 9 or 10).

**Comment by dengliming** [comment]
> @dengliming Thanks for the quick turnaround!
> Looks great!
> Please make sure you can build the agent using `./mvnw clean package` properly (with JDK 9 or 10).
I can build with JDK 9 local environment.

**Comment by eyalkoren** [comment]
@dengliming there is an inconsistency with the Jackson version brought by the Redisson version you used and the one used by the MockReporter. There are different solutions for that, but I opted for using the latest Redisson version for testing, which required some additional changes.
Another required change was the auto-created configuration.asciidoc file.
See the PR I did on this PR's branch and if it makes sense, please merge it.

In addition, it would be best if you knew exactly which Redisson versions we support with this PR, for other users to know. What is the earliest 3.x version supported?
Are 2.x versions supported? If not, please make sure our instrumentation does not break them.

Thanks again!

**Comment by dengliming** [comment]
@eyalkoren I can try to test all the version local to confirm that which version is really supported.

**Comment by dengliming** [comment]
@eyalkoren I've been testing it locally. The results are as follows.
3.0.0~3.12.3 (supported)
2.1.5~2.15.2 (supported)
And I added declaresMethod matcher "getChannel" for class org.redisson.client.RedisConnection. This method only exist in version: 2.1.5 or higher. So our instrumentation does not break the older version(I think).

**Comment by eyalkoren** [comment]
@dengliming great job on the version verification, thanks a lot!!
We could potentially still support older versions and not have the address for those, but since 2.1.5 is available since Nov 2015, this is more than enough.
Running the tests to see that they now pass.

File: docs/supported-technologies.asciidoc [hunk 3]
```diff
@@ -187,6 +187,11 @@ Other Servlet 3+ compliant servers will most likely work as well.
 |The agent creates spans for interactions with the Lettuce client.
 |1.13.0
 
+|Redis Redisson
+|3.x
```
Comment (eyalkoren) [inline docs/supported-technologies.asciidoc]:
```suggestion
|2.1.5+
```

File: docs/supported-technologies.asciidoc [hunk 4]
```diff
@@ -187,6 +187,11 @@ Other Servlet 3+ compliant servers will most likely work as well.
...
+|The agent creates spans for interactions with the Redisson client.
+|1.13.1
```
Comment (eyalkoren) [inline docs/supported-technologies.asciidoc]:
```suggestion
|1.15.0
```

File: docs/supported-technologies.asciidoc (hunk 3 above)
Comment (dengliming) [reply docs/supported-technologies.asciidoc]: Thanks.

**Comment by dengliming** [comment]
Sorry for the late reply. I'm a little busy today, I will CHANGELOG.asciidoc now.

**Comment by eyalkoren** [comment]
@dengliming Don't change the CHANGELOG just yet, we are probably releasing today or tomorrow (without this PR) so the entire section will move

**Comment by dengliming** [comment]
> @dengliming Don't change the CHANGELOG just yet, we are probably releasing today or tomorrow (without this PR) so the entire section will move

OK. Thanks

**Comment by eyalkoren** [comment]
@dengliming You can now merge from master, where I already prepared a [1.15.0 section](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc#features) in `CHANGELOG.asciidoc`, and add it as a new feature.

**Comment by dengliming** [comment]
@eyalkoren Please see co.elastic.apm.agent.profiler.asyncprofiler.JfrParserTest#name
this line：
`String filePath = URLDecoder.decode(url.getPath(), StandardCharsets.UTF_8);`
 StandardCharsets.UTF_8 not return string in JDK9.0.4

**Comment by eyalkoren** [comment]
Hmmm, thanks! The problem is that the overload of `java.net.URLDecoder#decode()` that accepts a `Charset` was only introduced in 10... 
Thanks!

**Comment by dengliming** [comment]
@eyalkoren One small problem is that after pr is merged, I should be a contributor to this project, But actually not, How does github calculate code commit？

**Comment by eyalkoren** [comment]
You are absolutely right! Let me see how we can fix this

**Comment by eyalkoren** [comment]
@dengliming I opened a GitHub support ticket, will let you know if there's anything to be done on your part to resolve this. Sorry for the inconvenience.

**Comment by dengliming** [comment]
Get it, Thanks.

**Comment by eyalkoren** [comment]
@dengliming This is the response I got from the GH support ticket I opened with regard to the code credit problem:
```
The behavior you’ve described and experienced was the result of an intentional 
change that our team recently deployed. Given the reaction and feedback from 
open source communities about the problems this has caused, as well as an 
identified bug within the squash and merge logic, a fix was deployed on 6 March 
to revert this change. Unfortunately, this fix will not correct existing attributions 
for pull requests that were merged prior to the deployed change.
```
We will figure something and make this right, just wanted to let you know we are on it.

**Comment by dengliming** [comment]
@eyalkoren Thank you so much, you are so kind, Please let me know if there's anything need to be done on my part to resolve this.

**Comment by felixbarny** [comment]
maybe we could revert and re-apply this PR?

**Comment by dengliming** [comment]
I can resubmit a pr if you want to do this.

**Comment by eyalkoren** [comment]
@dengliming We currently don't have a way to properly restore the original commits (with you as the author) because the branch is deleted. Please either restore the branch, or create a new PR where the Redisson plugin is deleted and re-added (in two separate commits) with you as author. Please just make sure it ends up in the exact current state.
Thanks and sorry for the inconvenience.

**Comment by dengliming** [comment]
@eyalkoren I really appreciate it, Let's forgot it and keep it that way current state. No need to do anything more.  I will contribute other feature next time.  Thanks again. ^_^

**Pull request merged**